torch_xla/csrc/tensor.cpp

```diff
@@ -265,7 +265,12 @@ XLATensor XLATensor::detach() const {
   XLATensor detached = XLATensor::clone(*this);
   detached.detach_();
   return detached;
-};
+}
+
+bool XLATensor::IsUndefined() const {
```

Comment:
How do we end up creating tensors like these?
The bridge::CreateXlaTensor() checks if tensor is defined.
I don't like the have an undefined XLA tensor, having a Data member, with an ATEN tensor in it, which is undefined.
An undefined XLA tensor should be a tensor with Data which is nullptr.
This is why there is an is_null() API, and the default ctor is "defaulted" and public.

Reply:
We create them from undefined ATen tensors. Since we can already represent ATen tensors in the `Data`, I didn't want to add yet another quirk of having data as `nullptr` being undefined. The main issue I have with it is that null data should mean empty tensor, not undefined -- not being to distinguish the two is weird. Therefore, if we want to represent both empty and undefined tensors, we'll need a singleton Data to represent undefined.

Reply:
But why do we want to have a Data for it?
You don't have to look far (besides an obvious unique_ptr/shared_ptr analogy), to see how "undefined tensors" are implemented.
Just look at at::Tensor default ctor, and you will notice it is exactly the same I am suggesting.
A data structure with an weak_ptr which is nullptr.

Reply:
The `impl_` in `at::Tensor` has `UndefinedTensorImpl` as a null type, which in turn only has one instance, its private `_singleton` field. `UndefinedTensorImpl` is also a `TensorImpl`. The closest we have to `TensorImpl` in `XLATensor` is `Data`. I don't fully understand your suggestion here, but we could introduce an `XLATensorImpl` to do similar things.

Reply:
Done.